beaker-vue/src/components/panels/PreviewPanel.vue

```diff
@@ -0,0 +1,379 @@
+        document.querySelector('body').removeEventListener('mouseup', this);
+    };
+    document.querySelector('body').addEventListener('mousemove', resize);
+    document.querySelector('body').addEventListener('mouseup', unhookResize);
+}
+
+type PreviewCategory =
+    | 'image' 
+    | 'csv' 
+    | 'tsv'
+    | 'plaintext' 
+    | 'pdf'
+    | 'excel'
+
+// text/csv should have more priority than text/?? or octet-stream for rich matching.
+const mimeCategory = (mimetype: string): PreviewCategory => {
+    if (mimetype.startsWith('image/')) {
+        return 'image'
+    }
+    if (mimetype === 'application/pdf') {
+        return 'pdf'
+    }
+    if (mimetype === 'text/csv') {
+        return 'csv'
+    }
+    if (mimetype === 'text/tsv') {
+        return 'tsv'
+    }
+    // todo: is this comprehensive?
+    if (mimetype === 'application/vnd.ms-excel' || 
+        mimetype === 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet') 
+    {
+        return 'excel'
+    }
+    // application/octet-stream is a common plaintext unknown
+    return 'plaintext'
+}
+
+type PreviewConfig = {
+    overridesToolbar?: boolean;
+    hasRawToggle?: boolean;
+}
+const mimetypeConfig: {[key in string]: PreviewConfig} = {
+    "application/pdf": {
+        overridesToolbar: true
+    },
+    "text/csv": {
+        hasRawToggle: true
+    },
+    "text/tsv": {
+        hasRawToggle: true
+    }
+}
+
+type PreviewContents = {
+    isLoading: boolean;
+    contents?: any;
+    contentLength: number;
+}
+const contents = ref<PreviewContents>({
+    isLoading: false,
+    contentLength: -1,
+});
+// should be in sync with isLoading, with false->true state change delayed by 500ms
+// (500ms being a dummy value for "when would a user want to interrupt a preview operation")
+const showPreviewInterrupt = ref(false);
+const previewInterruptTime = 500;
+
+const decoder = new TextDecoder()
+
+// max size in bytes to preview (10 MB for now)
+const maxSize = 10 * 1000000;
+
+// drag bar width in px
+const dragWidth = ref(5);
+// default pdfjs canvas plus dragbar size
+const width = ref(614 + dragWidth.value);
+const minWidth = 260;
+
+const props = defineProps([
+    'url', 
+    'mimetype'
+])
+
+const visible = defineModel<boolean>();
+
+const mime = computed(() => props.mimetype ?? fallbackMime.value ?? '')
+
+// contents as what is most useful by guessing at typing 
+let contentsWrapper = computed(() => {
+    const data = contents.value?.contents;
+    if (mime.value.startsWith('image/') || mime.value === 'application/pdf') {
+        return "<loaded elsewhere via url>";
+    }
+    if (!data) {
+        return "";
+    }
+    // even if data exists, handle 0-byte files.
+    if (data.length === 0) {
+        return "<file is 0 bytes>"
+    }
+
+    // sheetJS can injest Uint8Array without decoding or re-encoding
+    // catch-all for ms-office filetypes (todo: does anything else use vnd.*)?
+    if (mime.value.startsWith('application/vnd.')) {
```

Comment:
i think making this explicit is really good rather than just implicit from "expecting" it to be used via renderers.ts filtering mimetypes. handling it here should be way less hack-prone and ideally properly typed - e.g. expected data formats for plaintext tables vs document/binary formats, not relying on just mimetype.